### ISSUE TITLE
Cache digest and scoping

### ISSUE THREAD

**Issue by bryancall** [httpwg/http-extensions#216]
In the draft it is unclear what the scope the digest will cover.  Is it scoped to the hostname of the stream (authority header)?  Is the scope limited to the resources referenced in the stream and the client would have to keep a data structure around for every page?

There are a few cases where I can see issues with scoping on hostnames:
1. If a user goes to www.example.com, you might want to expand the cache digest to all of the resources for example.com.
2. The first point doesn't make sense for a site like Tumblr where there are many blogs for *.tumblr.com and creating a digest might become large.  Having a limit on the number of items in the digest and having the most used items in the cache digest might help.
3. There are also issues for sites where there are multiple domain names referenced by the stream and being served by the same connection.  This will happen with finance.yahoo.com where it has s.yimg.com resources and they are served by the same server.

For reference:
"A CACHE_DIGEST frame can be sent from a client to a server on any stream in the "open" state, and conveys a digest of the contents of the client's cache for associated stream."

**Comment by kazuho** [comment]
Thank you for your comment.

> In the draft it is unclear what the scope the digest will cover. Is it scoped to the hostname of the stream (authority header)?

In the current draft, the scope is defined in terms of _origin_, as:

```
   o  *COMPLETE* (0x2): When set, indicates that the currently valid set
      of cache digests held by the server constitutes a complete
      representation of the cache's state regarding that origin
(snip)
   When generating CACHE_DIGEST, a client MUST NOT include cached
   responses whose URLs do not share origins [RFC6454] with the request
   of the stream that the frame is sent upon.
```

> Is the scope limited to the resources referenced in the stream and the client would have to keep a data structure around for every page?

No. The server is expected to retain the digests sent by the client for the lifetime of the connection (or until it receives a `CACHE_DIGEST` frame with `RESET` flag set).

> There are a few cases where I can see issues with scoping on hostnames:

I think your examples illustrate a very good use-case that we might better cover.

The specifications in current draft forbids sending a digest of a _different_ origin than that specified in the `:authority` header.

It _might_ be a good idea to add a optional field to the frame that could be used to store the scope (in terms of origin) of the digests. Clients can decide whether or not to send digests covering multiple origins, depending on how the digest would become.

**Comment by bryancall** [comment]
Just to rephrase to make sure I am understanding this correctly.  If a request is make on a stream with the header {{:authority: www.yahoo.com}}, then only the resources in the cache digest would be from the hostname www.yahoo.com.  Is this understanding correct?

**Comment by martinthomson** [comment]
Note that you don't lose anything by including resources for `google.com` in the same cache-digest, but you do lose efficiency.

**Comment by kazuho** [comment]
@bryancall 
> If a request is made on a stream with the header `:authority: www.yahoo.com`, then only the resources in the cache digest would be from the hostname www.yahoo.com. Is this understanding correct?

Yes. That is what the current draft states (or tries to state).

**Comment by tombergan** [comment]
> When generating CACHE_DIGEST, a client MUST NOT include cached responses whose URLs do not share origins [RFC6454] with the request of the stream that the frame is sent upon.

I'm curious, is there an explanation somewhere why this design was chosen? There seems to be a trade-off between *coverage* (maximizing the amount of useful URLs in the digest) and *privacy* (minimizing the privacy leakages relative to HTTP/2 as it stands currently). As @bryancall points out, the current proposal has suboptimal coverage. To illustrate, here are two alternate strategies:

*Include all URLs for which the server is authoritative*: This is suggested by @bryancall. I believe this strategy maximizes coverage: a server can push any URL for which it is authoritative, and this strategy includes all such URLs in the digest. However, there is an increased chance for privacy leakage. Say a server is authoritative on a.com and b.com, and some user makes requests for a.com. If the cache digest includes URLs from b.com, the server can learn of those URLs even if the client never makes any requests for b.com.

*Tie the digest to cookies*: This idea is inspired by @kazuho's [H2O](https://h2o.examp1e.net/configure/http2_directives.html#http2-casper), which puts the cache digest in cookies. We can't actually store the digest in cookies (it's too inefficient), but if we designed the digest in such a way that it *could* be stored in cookies, at least in theory, then we could argue that cache digests do not increase the chance for privacy leakage relative to current H2. For example, suppose a URL can be included in a cache digest only if the following conditions hold:

* The server is authoritative for that URL.
* The client is making a request for origin https:x.y.foo.com, the URL's origin is X, and the client has a cookie that could be sent with requests for https:x.y.foo.com *and* with requests for X.

I believe this is roughly equivalent to the current proposal if the client does not have any domain-level cookies. However, if the client has a cookie for ".example.com" and the client makes a request for "www.example.com", the cache digest can include URLs from "cdn.example.com". Note that example #3 from @bryancall's initial comment would not work: on a request for "finance.yahoo.com", the digest would not include URLs from "s.yimg.com", although if the server is authoritative for both domains, the client may later make requests for "s.yimg.com", at which point the cache digest could be updated to include URLs on that origin as well.

> CACHE_DIGEST allows such cache-based fingerprinting to become passive, since it allows the server to discover the state of the client's cache without any visible change in server behaviour. As a result, clients MUST mitigate for this threat when the user attempts to remove identifiers (e.g., "clearing cookies").  This could be achieved in a number of ways; for example: by clearing the cache, ...

Expanding on the above design, we could view the cache digest as an internally-maintained cookie. When the browser makes a request to origin X, the browser creates a cookie with the domain set to host(X). If X is https, the cookie is Secure, otherwise it is HttpOnly. From this point forward, each time the browser caches a URL from origin X, it adds that URL to X's cache digest cookie. If a URL is evicted from the cache, it is removed from the cache digest cookie. If a page from origin X sets a cookie on domain(X), then cache cookies for all hosts Y where domain(Y) = domain(X) are linked, as it may be possible to send union(cookie(X),cookie(Y)) when an H2 connection is opened to either host. Note that the cache cookies for X and Y are linked *but not merged* as the external server may not be authoritative for both hosts -- this needs to be checked for each H2 connection.

Except for being created automatically, and being sent in CACHE_DIGEST rather than Cookie, cache digest cookies behave exactly like other cookies. They are cleared when the user asks for cookies to be cleared. They are disabled when the user disables cookies. They are cleared when the browser is closed when the browser was opened in Incognito mode.

There may be concerns about the following case. Suppose all hosts under `.aws.com` are served by the same cloud frontend, which terminates the H2 connection, but each host on this domain is served by a separate backend server. Now suppose `evil.aws.com` sets a cookie for `.aws.com`. Can `evil.aws.com` can spy on a user's from `alice.aws.com`? The answer is yes if the client sends a single cache digest that covers both hosts. I see two ways to prevent this:

1. Although a client may proactively send multiple cache digests, following the above rules, each digest summarizes URLs for a single origin only. (The cloud frontend would be trusted to not forward the digest for `alice.aws.com` to `evil.aws.com` -- this doesn't seem like a problem, since the frontend is already trusted to partition their traffic.) This may use extra bytes by requiring separate-origin digests to be sent in separate CACHE_DIGEST frames.

2. Cache digests from multiple origins can be merged, but the client will merge digests for `evil.aws.com` and `alice.aws.com` only if *both* origins have set a domain cookie for `.aws.com`. This potentially saves bytes by allowing one cache digest to cover the entire connection.

WDYT?

> When generating CACHE_DIGEST ... with the request of the stream that the frame is sent upon.

Relatedly, I'm a bit confused with the last part of that sentence. I would expect CACHE_DIGEST to apply to the entire connection (StreamID = 0x0) rather than an individual stream. Why not send CACHE_DIGESTS on stream 0x0 and include an explicit scope (authority) in the CACHE_DIGEST frame? This costs a few extra bytes but will (a) eliminate a race between receiving the first request and the digest, and (b) make it more obvious *which* authority the digest applies to.

**Comment by tombergan** [comment]
tl;dr of the above comment: There is a trade-off between coverage and privacy. Cache digests should not worsen the privacy problem compared to current H2. If cache digests are reducible to cookies, and managed as cookies, then we can argue that the privacy problem hasn't worsened. Further, if cookies are shared across hosts on the same domain, then cache digests can also be shared on the same domain, improving coverage.

**Comment by mnot** [comment]
@bryancall - the scope is the origin, which is a bit more than the `:authority` -- it's the tuple of `(scheme, hostname, port)`.

@tombergan and @bryancall - I think that tying cache digest to cookie scopes would be a very hard sell. All new mechanisms in the Web use origins as a default administrative boundary, and doing anything scoped bigger than that is going to invite a lot of attention and controversy.

Having said that, there's nothing that stops a client from sending multiple cache digests on a connection, one for each origin that the connection is authoritative for. If this needs to be clarified in the draft we can certainly do that.

**Comment by kazuho** [comment]
@mnot 
> Having said that, there's nothing that stops a client from sending multiple cache digests on a connection, one for each origin that the connection is authoritative for. If this needs to be clarified in the draft we can certainly do that.

I think CACHE_DIGEST frame should designate its scope explicitly, considering the fact that drafts that changes the scope of the authority exists (e.g. ORIGIN frame , secondary certificates), and that we have `COMPLETE` flag (that indicates a digest covering all the authorties have been sent).

Otherwise, there'd be a race condition.

**Comment by tombergan** [comment]
@mnot
> I think that tying cache digest to cookie scopes would be a very hard sell

FWIW, Section 4 recognizes the potential privacy concerns and tries to alleviate these concerns by tying cache digests to cookies, but in a weak way:

> As a result, clients MUST mitigate for this threat when the user attempts to remove identifiers (e.g., "clearing cookies") ... TODO: discuss how effective the suggested mitigations actually would be.

There were two parts to my suggestion (I didn't explain that well, sorry). The main suggestion: I think you can alleviate privacy concerns by showing that a server can already implement cache digests using cookies. For example, a server could send the following cookie with every response:

```plain
Name:    (URL, Validator)
Value:   "1"
Path:    "/"
Domain:  hostip(URL)                 // but see below
Expires: never                       // but see below
Secure:  true if scheme(URL)="https" // always true when H2 is https only
```

The RFC could say that the client should automatically create the above cookie for every response it receives from an origin. This cookie is treated exactly like every other cookie except that: (1) it is created automatically, (2) it expires when the response is evicted from cache, (3) it is sent in a cache digest, rather than the Cookie header field, (4) it is an internal cookie not available via the normal APIs (such as JS), (5) it does not count towards browser cookie limits, and (6) the Domain field is an exact-match (as in same-origin checking) rather than a suffix match as for normal cookies.

This exposition makes it obvious exactly where the privacy concerns are -- there are privacy concerns whenever cache digest cookies are treated differently than normal cookies. I don't think any of the above six differences are significant from a privacy perspective. (The only difference that might be significant is (5), but we could argue that a server could work around these limits in various ways, e.g., by using JS code to merge multiple cookies into a digest.) Some of the differences, like (6), are actually _stronger_ from a privacy perspective.

This exposition also makes it obvious where implementations might accidentally introduce privacy concerns. For example, if a browser is configured to reject third-party cookies, then it should _also_ reject third-party cache digest cookies, such as when an image from `foo.com` is used on a page from `bar.com`. Another example: if a browser partitions "incognito mode" cookies into a separate cookie store, it should do the same for cache digest cookies.

Lastly, my second (more controversial) suggestion is that the above exposition provides a path towards solving some of the problems raised by @bryancall. Namely, if `www.foo.com` has already stated willingness to share cookies with `foo.com`, then it _may_ be reasonable for `www.foo.com` to share cache digest cookies with `foo.com`. Though, on second thought, it's probably more reasonable to have an explicit signaling mechanism, rather than an implicit signaling mechanism. A Cache-Control extension may do the trick, as suggested in [this thread](https://lists.w3.org/Archives/Public/ietf-http-wg/2016JulSep/0520.html).

**Comment by kazuho** [comment]
@tombergan 
> Cache digests should not worsen the privacy problem compared to current H2. If cache digests are reducible to cookies, and managed as cookies, then we can argue that the privacy problem hasn't worsened.

I agree that cache digests should not worsen the privacy problem.

However, I will also argue that cache digests _in current form_ does not worsen the privacy problem since it is something that can be and is implemented using existing technology (stated below), and that there is no need to try to adopt the cookie model which would have issues as @mnot points out.

Cache digest can be (and already is) implemented using Service Worker. The only differences between a Service Worker-based implementation and the HTTP/2 frame-based implementation is that the frame-based implementation is more efficient (in HTTP/2 frame-based approach, the digest needs to be sent only once per connection, whereas for Service Worker-based or cookie-based implementation, a client needs to send a digest per every request), and that the digest more accurately reflects the state of the cache (Service Worker cache is located in front of the web browser cache, so a digest of Service Worker cache does not reflect what exists in the browser's cache).

**Comment by tombergan** [comment]
We completely agree. I was making a writing suggestion: if you can write the RFC in such a way that it's obvious that the proposed cache digest frame is equivalent to cookies, then you can more easily argue that you're not introducing any privacy concerns. [1] My specific suggestion is for the browser to create a "cookie" each time it caches a response, where that cookie is treated like any other cookie with a few small changes (see the list of six items in my prior comment).

Rephrasing the explanation in terms of cookies also makes it obvious where privacy holes may arise. For example, if a browser supports "normal" and "incognito" modes that use different cookie jars but the same cache, the browser needs to attach the "cache cookies" to the appropriate cookie jar rather than to the cache, otherwise there may be information leakage across browser modes.

Does that sound reasonable? Contrast this to [Section 4](https://tools.ietf.org/html/draft-ietf-httpbis-cache-digest-01#section-4), which is currently kind of hand-wavy.

[1] Even if you rephrase the explanation in terms of cookies, privacy concerns are still likely to arise since the cookies are created automatically by the browser rather than manually by the server, but that is a separate issue.

**Comment by kazuho** [comment]
@tombergan 
> I realize cookies are kind of weird because they don't exactly follow the same-origin policy. Maybe a better analogy would be local storage?

Thank you for the suggestion.

I agree that it would be worthwhile to suggest similarity to existing mechanism in Security Considerations. Having that would help the readers understand the issues.

I assume that the existing mechanism we would refer to to be Service Worker (and its cache). That is natural since we are likely to have an appendix explaining how to implement Cache Digests using Service Workers (see #256).
